org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/WorkspaceDiagnosticsHandler.java

```diff
@@ -127,6 +128,12 @@ public boolean visit(IResourceDelta delta) throws CoreException {
 		// Check if it is a Java ...
 		if (JavaCore.isJavaLikeFileName(file.getName())) {
 			ICompilationUnit cu = (ICompilationUnit) JavaCore.create(file);
+			// Clear the diagnostics for the resource not on the classpath
```

Comment:
Wouldn't it make more sense go listen for classpath changes on the java model and remove diagnostics when a folder is removed from the classpath?

Reply:
You can start from JavaCore#addElementChangedListener() to learn about Java model deltas.

Reply:
@tsmaeder Thanks for the info. Have a try on JavaCore#addElementChangedListener(). 
Considering the scenario below:
```
create A,java file in source folder
create newfolder, add as source folder
create B.java file in new source folder
make A extend B
remove newfolder from source path
```
Below is the execution flow:
```
Invoke the API JavaProject.setRawClasspath() to update the project's classpath.
-> IElementChangedListener receives .classpath Resource Delta (Remove From Classpath Event).
-> AutoBuildJob is triggered automatically.
-> IElementChangedListener receives the changed model. (Include A.java).
-> WorkspaceDiagnosticsHandler (because it implements IResourceChangeListener) will also receive the changed resource. (Include A.java).
```
Both listeners do the same thing. Their handle logic are similar, importing a new listener will cause duplicated logic to handle the same work. 

So i prefer to use WorkspaceDiagnosticsHandler to refresh diagnostics.